Refatoração No Zeus Aqui irá ter NF-e 4.00 (Ainda não tem). Grandes melhorias.

robertorp opened the pull request:
Refatoração em toda estrutura.
Oque mudou?
Agora temos somente uma unica DLL que se chama DFe


Exemplo de chamada

```
var mdfeFacade = new MDFeFacade(config, certificado);

mdfeFacade.EnviarLote(idLote, mdfe);
```

Atualmente se encontra padronizado o CT-e e MDF-e.

- [ ] Compatibilidade com .Net Standard 2.0
- [x] CT-e Classes de WSDL preparadas para compatibilidade com .Net Standard 2.0
- [ ] NF-e Classes de WSDL preparadas para compatibilidade com .Net Standard 2.0
- [x] MDF-e Classes de WSDL preparadas para compatibilidade com .Net Standard 2.0


De fato não foi testado em um SO diferente de windows, na teoria esta tudo ok, na pratica não sei ainda. Quando alguém implementar algo para testar no linux ou quando eu terminar meu emissor online testarei. Mas se alguém tiver a oportunidade de testar antes em algum linux/mac agradeço. Obrigado.


Projeto DFeStandard Criado, nele já temos o CT-e/MDF-e totalmente compatível.


- [ ] Trocar todas combinações de caminhos por ``` Path.Combine``` para melhor compatibilidade com .net standard


- [X] Adicionado um primeiro projeto de Testes unitários 


CT-e OS
- [x] Ajustes nas classes de geração do CT-e
- [x] Criar classes separadas para CT-e OS
- [x] Enviar CT-e OS

Estou Iniciando a NF-e , a NF-e já vou padronizar ela com NF-e 4.0!

Andamento das alterações do projeto NF-e com inclusão de NF-e 4.0 
- [ ] Verificar questão de reset do certificado digital de acordo com o commit https://github.com/adeniltonbs/Zeus.Net.NFe.NFCe/commit/78cb7dbaf437925bacf9836e3b7fc54cb14efda8
- [x] Status do Serviço
- [ ] Inutilizar Númeração
- [ ] Carta Correção
- [ ] Cancelar 
- [ ] Autorizacao
- [ ] Consulta Recibo
- [ ] Manifestação Destinatário
- [ ] Evento EPEC
- [ ] Consulta Protocolo
- [ ] Consulta Cadastro
- [ ] Distribuição
- [ ] Remover NfeDownloadNf (Foi descontinuado pela SEFAZ)
- [ ] Administrar CSC NFC-e
- [ ] Revisão

Impressões 
- [ ] Impressões NF-e
- [ ] Impressões CT-e
- [ ] Impressões NFC-e
- [ ] Impressões MDF-e


- [ ] Publicar NUGET

Estou deixando aqui para todos verem como esta ficando.
Se alguém quiser dar contribuir que seja com ajuda ou dando apenas ideias para estar melhorando fiquem a vontade. 
Previsão de conclusão: 01/01/2018

Algumas observações


NT NT_2016_002_v1.31.pdf final da página 11
Nota: Os webservices abaixo foram desativados, conforme definido na NT2014.002, v1.02b:
4.11.6 Validação das informações de controle da chamada ao Web Service (NfeConsultaDest);
4.12.6 Validação das informações de controle da chamada ao Web Service (NfeDownloadNF);

robertorp committed:
unificado os três projetos
faltando refatorações

robertorp committed:
Refatorado EnviarLote MDFe

robertorp committed:
Feito Consulta Lote

robertorp committed:
Feito consulta
Feito StatusConsulta

robertorp committed:
Feito MDFeCancelar

robertorp committed:
Incluido MDFeEncerrar
Incluido MDFeIncluirCondutor

robertorp committed:
Criado um DFeConfig padrão

robertorp committed:
ajustes para melhor uso da lib

robertorp committed:
Mudando namespaces padronizando mdf-e

robertorp committed:
tudo ajustado para nf-e 3.0
faltando a 2.0

robertorp committed:
Fazendo melhorias para mdf-e 1.0

robertorp committed:
Refatoração em muitos pontos do mdf-e

robertorp committed:
Criado classes que manipularão o certificado digital

robertorp committed:
Aplicado nova configuracao

robertorp committed:
Feito novo esquema de salvar xmls

robertorp committed:
Feito base para cache do certificado digital

robertorp committed:
Concluido cache de certificado digital

robertorp committed:
Ajustes de namespaces

robertorp committed:
Trocado namespace de documentos

robertorp committed:
Melhorias no MdfeAppTeste

robertorp committed:
Refatorado servicos do cte

robertorp committed:
Ajustado todas extencoes CTE

robertorp committed:
ajustando eventos

robertorp committed:
ajustes em evento de ct-e

robertorp committed:
refatorado namespaces de classes do ct-e

robertorp committed:
Refatorado namespaces wsdl cte

robertorp committed:
Criado facade
Trocado local de ControleCache

robertorp committed:
Refatorado métodos de eventos CTe

robertorp committed:
Adicionado Config é CertificadoDigital em classes do cte

robertorp committed:
Ajustes no CTeTesteModel trocado implementação direta para Facade

robertorp committed:
Resolvido problema de organização de pastas

robertorp committed:
Ajustes da refatoração

robertorp committed:
Correção no Cancelamento de CT-e

marcosgerene commented:
Excelente, de cara isso mata duas insues "complexas" com louvor!

https://github.com/adeniltonbs/Zeus.Net.NFe.NFCe/issues/495

https://github.com/adeniltonbs/Zeus.Net.NFe.NFCe/issues/490

Como conversamos anteriormente via whatsapp me comprometo a revisar todas impressões em fast reports.

----

Sobre o refatoring, acredito que a melhor opção para framework seja o .Net 4.5.2.

robertorp committed:
Refatorado namespaces NF-e

robertorp committed:
Todos projetos setados para .net 4.5.2

robertorp committed:
Adicionado TLS 1.2 AppTeste

robertorp committed:
Trocado Enum de VersaoServico agora ele é unico para os três projetos

robertorp committed:
Trocado TipoAmbiente para um padrão de todos documentos eletronicos

robertorp committed:
Ajuste em namespace
Feito Factorys para NF-e

robertorp committed:
Retirado xmlelement do docAnt

robertorp committed:
Adicionado licença no email builder

robertorp committed:
Adicionaod licença

robertorp committed:
Revert "Adicionaod licença"

This reverts commit f9b0b4539525b56ea7ff99825af80e27f913f3c5.

marcosgerene commented:
@robertorp e @adeniltonbs 

Acredito que não passa de uma bobeira, mas já que teremos uma reescrita desse tamanho e tal não convém mudar o nome atual do componente para "Zeus.DFe"

Acredito que ficaria bem mais "apresentável"

robertorp committed:
storage.ide

robertorp committed:
storage.ide

robertorp committed:
Deletado projetos não usados mais.

robertorp committed:
storage.ide

robertorp committed:
merge

robertorp committed:
Atualizado schemas NF-e

robertorp committed:
deletado nf-e classes

marcosgerene commented:
Boa noite @robertorp 

Sobre as impressões penso em algo assim

DFe.Impressao.Ferramenta.

Ex:
**DFe.Impressao.Nativo**
**DFe.Impressao.FastReports** (me comprometo a deixar este funcionando)
_DFe.Impressao.DBExpress_

Neste caso o desenvolvedor vai usar apenas 2 dlls, DFe.dll e DFe.Impressao.Ferramenta.dll

robertorp committed:
Refatorado método statusServico NF-e com NF-e 4.0

robertorp committed:
storage.ide

robertorp committed:
Alterado comportamento do enderEmit de MDF-e se setar em branco ele não é serializado.

robertorp committed:
sotrage.ide

robertorp committed:
merge

alexsandro-xpt commented:
@robertorp tem como contribuir para o projeto sem ter certificado A1 ou A3?

robertorp committed:
Ajustes CT-e

robertorp committed:
Add git ignore sotrage.ide

robertorp committed:
removendo pasta .vs/

robertorp committed:
git ignore

robertorp committed:
Renomado Solution retirado espaço

robertorp commented:
@alexsandro-xpt  Crie uma issue para sua duvida, aqui é somente para a refatoração.

robertorp committed:
merge master

robertorp committed:
Valida

robertorp committed:
merge

robertorp committed:
Deletado NFeTestes

robertorp committed:
Adicionado na tela do app teste nfeautorizacao 4.00

robertorp committed:
Criado classes para Enviar NF-e 4.00

robertorp committed:
Retirado tudo oque mistura de CT-e os (67) com o CT-e (57) ,
Criar novo namespace para CTeOS

robertorp committed:
Inicio de criação de classes para geração de xml CT-eOS (67)

robertorp committed:
Criado todas as classes para geração de ct-e OS

robertorp committed:
fazendo classes de emissão de cteos

robertorp committed:
Enviando CT-e OS

robertorp committed:
Salvando retorno do cteOS

robertorp committed:
adicionado vINSS para emissão do ct-e OS

robertorp committed:
Melhorado tela de apresentação do AppTeste CT-e

robertorp committed:
Feito classe para Fazer request na sefaz compativel com .net standard é .net framework

robertorp committed:
Fazendo SoapHttp Base

robertorp committed:
Fazendo novo modelo de request SOAP para as Sefazes
O novo modelo deve ter compatibilidade com .net Standard

robertorp committed:
Retirado appTeste console e adicionado testes de unidade

robertorp committed:
Melhorias na separação de classes para consumo de soap

robertorp committed:
Alterado Autorização de CT-e OS para nova comunicação feita para ser compativel com .net standard

robertorp committed:
Adicionado compatibilidade da classe CteRecepcao

robertorp committed:
Adicionado compatibilidade da classe CteRetRecepcao para .net standard

robertorp committed:
Adicionado compatibilidade da classe CteConsulta para .net standard

robertorp committed:
Adicionado compatibilidade da classe CTeInutilizacao para .net standard

robertorp committed:
Adicionado compatibilidade da classe CTeStatusConsulta para .net standard

robertorp committed:
Trocado namespace de Gerado para Comunicacao
Todas comunicações de WSDL do CT-e agora são feitas manualmente, deixando de lado o código gerado pelo comando wsdl no cmd do visual studio.

robertorp committed:
Adicionado compatibilidade da classe CteRecepcaoEvento para .net standard

robertorp committed:
Adicionado compatibilidade da classe MDFeConsNaoEnc para .net standard
Criado estrategia para MdfeTagCabecalho (Mensagem Soap)
Criado estrategia para MdfeTagCorpo (Mensagem Soap)

robertorp committed:
Adicionado compatibilidade da classe MDFeConsulta para .net standard

robertorp committed:
Adicionado compatibilidade da classe MDFeRecepcaoEvento para .net standard

robertorp committed:
Adicionado compatibilidade da classe MDFeRecepcao para .net standard

robertorp committed:
Adicionado compatibilidade da classe MDFeConsultaRecepcao para .net standard

robertorp committed:
Adicionado compatibilidade da classe MDFeStatusServico para .net standard

robertorp committed:
Trocado namespace de Gerado para Comunicacao
Todas comunicações de WSDL do MDF-e agora são feitas manualmente, deixando de lado o código gerado pelo comando wsdl no cmd do visual studio.

robertorp committed:
Iniciando novo NFceCSC para nova versão do Zeus

robertorp committed:
clear

robertorp committed:
Adicionado Projeto DFeStandard

robertorp committed:
Adicionado Referencia de System.Security.Cryptography.Xml no projeto DFeStandard

robertorp committed:
Adicionado Compatibilidade de CT-e/MDF-e com .net standard 2.0

robertorp committed:
Ajustes para .net standard/.net framework

robertorp committed:
Retirado validação de schema para .net core atravez de um boolean setando como true e retirado a validação de shcema interna.
Motivo é que não achei uma solução funcional para .net core a que está ira executar mas irá dar falha.

robertorp committed:
NF-e Autorização Refatoraçao para novo modelo de negocio, faltando adicionar Layout Novo NF-e 4.0

robertorp committed:
Ajuste no appTeste

robertorp commented:
Utilizando sistema operacional Windows, com docker Linux.

Criei um projeto .net core Web atraves do VS 2017, ele me da a opção de docker, mandei criar com docker linux.  (Lembre-se de instalar o docker antes no windows)

O próprio visual studio fez meu docker linux rodando no Hyper-V

nas configurações precisei colocar 

```ProtocoloDeSeguranca = SecurityProtocolType.Tls12 | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls```

é tirar a validação do certificado digital 

Antes de executar um 
```
var facade = new MDFeFacade(configMdfe, configCert);

ServicePointManager.ServerCertificateValidationCallback = delegate { return true; };

var retorno = facade.StatusConsulta();
```

Feito isso, funcionou normalmente no Docker Linux.

robertorp committed:
mdf-e consulta status trocado concateção de pastas por combine path

robertorp committed:
Alterações de diretorios nos Links do projeto DFeStandard

robertorp committed:
Adicionado Path.Combine para MDF-e é CT-e

robertorp committed:
Remanejo de pastas

robertorp committed:
Adicionado uma pasta raiz "src"

robertorp committed:
Implementado Certificado Digital em Bytes de acordo com merge https://github.com/adeniltonbs/Zeus.Net.NFe.NFCe/pull/540

robertorp committed:
merge

robertorp committed:
Adicionado Teste novamente no appteste de certificado digital em bytes
equivalente ao pull https://github.com/adeniltonbs/Zeus.Net.NFe.NFCe/pull/540

robertorp committed:
Criado Testes para StatusServico MDF-e

robertorp committed:
cep enderTomaOs não e obrigatorio

robertorp committed:
Correções de layout xml

robertorp committed:
merge

robertorp committed:
Removido impressões para melhor desenvolvimento.
Quando tiver tudo redondo, pego os projetos de impressões novamente do master e coloco os mesmo aqui.

robertorp committed:
Fazendo melhorias

robertorp committed:
Adicionado Eventos no envio de CT-e OS

robertorp committed:
alterado nome de classes do antigo projeto para não ter problemas ao referenciar as mesma utilizando dll zeus e dll dfe ao mesmo tampo

robertorp committed:
Retirado memory leak
https://github.com/ZeusAutomacao/DFe.NET/issues/610

robertorp committed:
Adicionado verificações de isAutorizado e isDenegada e isCancelada no retorno da consulta situação ct-e

robertorp committed:
adicionado condição para identificar rejeição

robertorp committed:
Feito evento para antes de consulta

robertorp committed:
adicionado metodos auxiliares para cte os

apbertoletti commented:
Roberto, só me tira uma duvida...

Estamos pensando em integra o Zeus aqui no nosso ERP, mas até agora não havíamos "batido o martelo" pois estávamos aguardando a liberação da NFe 4.0, pra tentar evitar retrabalho e assim aproveitar fazer tudo de uma vez.

Mas tendo em vista de que a nova NFe no Zeus não será pra agora, me veio a duvida: será compensa nós já integrarmos o Zeus do que já está hoje, ou seria melhor aguardar a NFe 4.0 mesmo? Pergunto isso pois não sei se forma de consumir a NFe 4.0 pelo Zeus seria muito diferente do que é hoje com a NFe 3.10?

Muito obrigado

robertorp commented:
@apbertoletti  Bom vai ser mantido duas versões do Zeus uma na versão antiga e essa na versão nova, então compensa sim você mexer no Zeus atualmente que tem a nf-e 3.10 posteriormente ele sera mantido também. (Ate mesmo meu erp esta no zeus antigo do zeus novo oque tem no meu erp e CT-e OS, o resto ta no zeus antigo, esse pull vai demorar ainda.. o zeus antigo vai ser mantido por muito tempo afinal ate mesmo eu utilizo ele) então compensa colocar no Zeus antigo você tera mais de ano tranquilo de manutenção nele.

apbertoletti commented:
Certo @robertorp, 

Só mais uma coisa que fiquei na duvida:

Na versão antiga do Zeus tb estará disponível a NFe 4.0 ? 

Obrigado pelos esclarecimentos

robertorp commented:
Na versão antiga do Zeus tb estará disponível a NFe 4.0 ? Sim

robertorp commented:
Ah sim entendi, sobre o xml e serialização não tenho problemas e nunca tive, e muito rapido a nativa do C# não vejo motivos para mudar, ao fazer manual igual fazemos no zeus tem a vantagem de escrevermos como quisermos até estou pensando em fazer algo diferente do xml seguir um padrão próprio.
Exemplo: NFeEmit ou Emit seguir padrões do C# . Mas vamos estudando as possibilidades.

robertorp committed:
Atualizado pacotes

robertorp committed:
Adicionado consulta a web service status de forma nativa na nf-e

robertorp committed:
clear

robertorp committed:
Adicionado tagcabecalhovazia

robertorp committed:
merge

robertorp committed:
Enderecador 4.00 urls adicionado

eliezeralmeida committed:
removido log na exception ao receber falha no envio da consulta do cte para sefaz

robertorp committed:
adicionado tags infresptec

robertorp committed:
remover acentuação

robertorp committed:
remover acentos

robertorp committed:
tratamento para rejeição 999

eliezeralmeida committed:
correção no CEP do Tomador para adicionar zeros a esquerda

robertorp committed:
cte os denegado

robertorp commented:
melhorias daqui serão transportadas para o master aos poucos!
